contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -0,0 +1,297 @@
+	function stakeTokensWithApproval(
+		address _sender,
+		uint256 _amount,
+		uint256 _restartStakeSchedule
+	) external onlyThisContract returns (uint256 lastSchedule, uint256 remainingAmount) {
+		(lastSchedule, remainingAmount) = _stakeTokens(_sender, _amount, _restartStakeSchedule);
+	}
+
+	/**
+	 * @notice Delegate votes from `msg.sender` which are locked until lockDate
+	 * to `delegatee`.
+	 * @param _delegatee The address to delegate votes to.
+	 * */
+	function delegate(address _delegatee) external onlyTokenOwner {
+		require(_delegatee != address(0), "delegatee address invalid");
+		uint256 stakingEndDate = endDate;
+		/// @dev Withdraw for each unlocked position.
+		/// @dev Don't change FOUR_WEEKS to TWO_WEEKS, a lot of vestings already deployed with FOUR_WEEKS
+		///		workaround found, but it doesn't work with TWO_WEEKS
+		for (uint256 i = startDate.add(cliff); i <= stakingEndDate; i += FOUR_WEEKS) {
+			staking.delegate(_delegatee, i);
+		}
+		emit VotesDelegated(msg.sender, _delegatee);
+	}
+
+	/**
+	 * @notice Withdraws unlocked tokens from the staking contract and
+	 * forwards them to an address specified by the token owner.
+	 * @param receiver The receiving address.
+	 * */
+	function withdrawTokens(address receiver) external onlyTokenOwner {
+		_withdrawTokens(receiver, false);
+	}
+
+	/**
+	 * @notice Collect dividends from fee sharing proxy.
+	 * @param _loanPoolToken The loan pool token address.
+	 * @param _maxCheckpoints Maximum number of checkpoints to be processed.
+	 * @param _receiver The receiver of tokens or msg.sender
+	 * */
+	function collectDividends(
+		address _loanPoolToken,
+		uint32 _maxCheckpoints,
+		address _receiver
+	) external onlyTokenOwner {
+		require(_receiver != address(0), "receiver address invalid");
+
+		/// @dev Invokes the fee sharing proxy.
+		feeSharingProxy.withdraw(_loanPoolToken, _maxCheckpoints, _receiver);
+
+		emit DividendsCollected(msg.sender, _loanPoolToken, _receiver, _maxCheckpoints);
+	}
+
+	/**
+	 * @notice Allows the owners to migrate the positions
+	 * to a new staking contract.
+	 * */
+	function migrateToNewStakingContract() external onlyOwners {
+		staking.migrateToNewStakingContract();
+		staking = Staking(staking.newStakingContract());
+		emit MigratedToNewStakingContract(msg.sender, address(staking));
+	}
+
+	/**
+	 * @notice Extends first year stakes for four year vesting contracts.
+	 * @dev Tokens are vested for 4 years. Since the max staking
+	 * period is 3 years and the tokens are unlocked only after the first year is
+	 * passed, hence, we extend the duration of staking for all unlocked tokens for the first
+	 * year by 3 years.
+	 * */
+	function extendStaking() external {
+		uint256 oneYear = startDate.add(52 weeks);
+		uint256[] memory dates;
+		uint96[] memory stakes;
+		(dates, stakes) = staking.getStakes(address(this));
+
+		for (uint256 i = 0; i < dates.length; i++) {
+			if ((dates[i] < block.timestamp) && (dates[i] <= oneYear) && (stakes[i] > 0)) {
```

Comment:
(dates[i] < block.timestamp) - we can't extend expiring stakes, so we need this check.
It's possible if `extendStaking` wasn't invoked or invocation failed for some vesting:
in this case tokens will be locked because we can't extend stake and can't withdraw stakes for first year (withdraw - startDate.add(52 weeks))

If stake for the first year expired we need to withdraw it and stake again to an appropriate date (dates[i].add(156 weeks)).

Reply:
@smbsp you're right, it will be possible to extend